python/src/uagents/communication.py

```diff
@@ -0,0 +1,268 @@
+
+    def add_envelope(self, envelope: Envelope, sync: bool = False):
+        self._envelopes.append((envelope, sync))
+
+    async def run(self):
+        while True:
+            for env, sync in self._envelopes:
+                try:
+                    await send_exchange_envelope(
+                        envelope=env,
+                        resolver=self._resolver,
+                        sync=sync,
+                    )  # how do we handle the MsgStatus response?
+                    self._envelopes.remove((env, sync))
+                except ValueError:
+                    continue
+            await asyncio.sleep(DEFAULT_DISPENSER_INTERVAL)
+
+
+async def dispatch_local_message(
+    sender: str,
+    destination: str,
+    schema_digest: str,
+    message: JsonStr,
+    session_id: uuid.UUID,
+) -> MsgStatus:
+    """Process a message locally."""
+    await dispatcher.dispatch(
+        sender=sender,
+        destination=destination,
+        schema_digest=schema_digest,
+        message=message,
+        session=session_id,
+    )
+    return MsgStatus(
+        status=DeliveryStatus.DELIVERED,
+        detail="Message dispatched locally",
+        destination=destination,
+        endpoint="",
+    )
+
+
+async def send_exchange_envelope(
+    envelope: Envelope,
+    resolver: Optional[Resolver] = None,
+    sync: bool = False,
+) -> Union[MsgStatus, Envelope]:
+    """
+    Method to send an exchange envelope.
+
+    Args:
+        envelope (Envelope): The envelope to send.
+        resolver (Optional[Resolver], optional): The resolver to use. Defaults to None.
+        sync (bool, optional): True if the message is synchronous. Defaults to False.
+
+    Returns:
+        Union[MsgStatus, Envelope]: Either the status of the message or the response envelope.
+    """
+    if resolver is None:
+        resolver = GlobalResolver()
+    _, endpoints = await resolver.resolve(envelope.target)
+    if len(endpoints) == 0:
+        LOGGER.error(
+            f"Unable to resolve destination endpoint for address {envelope.target}",
+        )
+        return MsgStatus(
+            status=DeliveryStatus.FAILED,
+            detail="Unable to resolve destination endpoint",
+            destination=envelope.target,
+            endpoint="",
+        )
+    headers = {"content-type": "application/json"}
+    if sync:
+        headers["x-uagents-connection"] = "sync"
+    for endpoint in endpoints:
+        try:
+            async with aiohttp.ClientSession() as session:
+                async with session.post(
+                    endpoint,
+                    headers=headers,
+                    data=envelope.json(),
+                ) as resp:
+                    success = resp.status == 200
+                    if success:
+                        if sync:
+                            return await dispatch_sync_response_envelope(
+                                Envelope.parse_obj(await resp.json())
+                            )
```

Comment:
This should only dispatch the envelope if there is an agent to dispatch to, which is not always the case for this standalone function (see https://github.com/fetchai/uAgents/blob/main/python/examples/08-local-network-interaction/sync_sender.py). Otherwise it should return the envelope itself.